--- madwifi-ng-r1686-20060715/ath/if_ath.c.orig	2006-09-26 23:14:40.000000000 +0200
+++ madwifi-ng-r1686-20060715/ath/if_ath.c	2006-09-26 23:15:27.421137500 +0200
@@ -4677,7 +4677,7 @@
 	while (an->an_uapsd_qdepth) {
 		bf = STAILQ_FIRST(&an->an_uapsd_q);
 		STAILQ_REMOVE_HEAD(&an->an_uapsd_q, bf_list);
-		bf->bf_desc->ds_link = (u_int32_t)NULL;
+		bf->bf_desc->ds_link = 0;
 
 		dev_kfree_skb_any(bf->bf_skb);
 		bf->bf_skb = NULL;
@@ -4693,7 +4693,7 @@
 	while (an->an_uapsd_overflowqdepth) {
 		bf = STAILQ_FIRST(&an->an_uapsd_overflowq);
 		STAILQ_REMOVE_HEAD(&an->an_uapsd_overflowq, bf_list);
-		bf->bf_desc->ds_link = (u_int32_t)NULL;
+		bf->bf_desc->ds_link = 0;
 
 		dev_kfree_skb_any(bf->bf_skb);
 		bf->bf_skb = NULL;
@@ -6241,7 +6241,7 @@
 	while (an->an_uapsd_qdepth) {
 		bf = STAILQ_FIRST(&an->an_uapsd_q);
 		STAILQ_REMOVE_HEAD(&an->an_uapsd_q, bf_list);
-		bf->bf_desc->ds_link = (u_int32_t)NULL;
+		bf->bf_desc->ds_link = 0;
 		txq = sc->sc_ac2q[bf->bf_skb->priority & 0x3];
 		ath_tx_txqaddbuf(sc, ni, txq, bf, bf->bf_desc, bf->bf_skb->len);
 		an->an_uapsd_qdepth--;
@@ -6250,7 +6250,7 @@
 	while (an->an_uapsd_overflowqdepth) {
 		bf = STAILQ_FIRST(&an->an_uapsd_overflowq);
 		STAILQ_REMOVE_HEAD(&an->an_uapsd_overflowq, bf_list);
-		bf->bf_desc->ds_link = (u_int32_t)NULL;
+		bf->bf_desc->ds_link = 0;
 		txq = sc->sc_ac2q[bf->bf_skb->priority & 0x3];
 		ath_tx_txqaddbuf(sc, ni, txq, bf, bf->bf_desc, bf->bf_skb->len);
 		an->an_uapsd_overflowqdepth--;
