--- madwifi-ng-r1757-20061020/ath/if_athvar.h~	2006-10-18 13:38:57.000000000 +0200
+++ madwifi-ng-r1757-20061020/ath/if_athvar.h	2006-10-21 02:35:04.000000000 +0200
@@ -189,7 +189,7 @@ static inline struct proc_dir_entry *PDE
 #define ATH_MAX_MTU     2290
 #define ATH_MIN_MTU     32  
 
-#define	ATH_RXBUF	40		/* number of RX buffers */
+#define	ATH_RXBUF	90		/* number of RX buffers */
 #define	ATH_TXBUF	200		/* number of TX buffers */
 
 #define	ATH_BCBUF	4		/* number of beacon buffers */
