--- madwifi-ng-r3556-20080421/Makefile.inc	2008-04-18 02:08:35.000000000 +0000
+++ madwifi-ng-r3556-20080421/Makefile.inc	2008-07-22 18:41:38.000000000 +0000
@@ -148,7 +148,7 @@
 TOOLS=  $(TOP)/tools 
 
 WARNINGS = -Werror -Wall
-COPTS+= $(WARNINGS)
+#COPTS+= $(WARNINGS)
 INCS=	-include $(TOP)/include/compat.h -I$(TOP)/include
 
 # TARGET defines the target platform architecture. It must match one of
